splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeSortJoinStrategy.java

```diff
@@ -72,4 +74,24 @@ public JoinStrategyType getJoinStrategyType() {
     public int maxCapacity(int userSpecifiedCapacity,int maxMemoryPerTable,double perRowUsage){
         return Integer.MAX_VALUE;
     }
+
+    private boolean correlatedSubqueryRestriction(Optimizer optimizer,
+                                                  OptimizablePredicateList predList,
+                                                  int joinType) throws StandardException {
+        if (optimizer.isForSpark()
+            && (joinType == LEFTOUTERJOIN || joinType == FULLOUTERJOIN)
```

Comment:
Why is the fix limited to outer joins?  Inner join also does a shuffle.  Do we need to hint non-native spark to see the problem for inner joins?

Reply:
In case of inner join, the subquery is not part of the join predicate itself and it is pulled as a restriction of a PRN on top of the join, as shown in the plan below, that's why it is OK, but I think your point is nevertheless valid, if, for some reason, the subquery becomes part of the join condition, then we will definitely run into this problem if we choose sort-merge (merge-sort?) join strategy. I apply this restriction for all types of joins.

```
explain SELECT * FROM 
t1
where a in (
select ALPHA.a from
t1 ALPHA  
INNER JOIN 
t2 BETA --splice-properties joinstrategy=sortmerge, useSpark=True
ON BETA.a = (SELECT MAX(A) FROM t2 SUBQ WHERE SUBQ.b=ALPHA.b) --spark-properties doNotFlatten=true
AND ALPHA.b=BETA.b
);

Cursor(n=16,rows=1,updateMode=READ_ONLY (1),engine=OLAP (query hint))
  ->  ScrollInsensitive(n=16,totalCost=8.011,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3)
    ->  ProjectRestrict(n=15,totalCost=4.001,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3,preds=[is not null(subq=14)])
      ->  Subquery(n=14,totalCost=10.682,outputRows=1,outputHeapSize=16 B,partitions=1,parallelTasks=3,correlated=false,expression=false,invariant=true)
        ->  ProjectRestrict(n=14,totalCost=8.008,outputRows=1,outputHeapSize=16 B,partitions=1,parallelTasks=3,preds=[(A[1:1] = ALPHA.A[13:1])])
          ->  ProjectRestrict(n=13,totalCost=8.008,outputRows=1,outputHeapSize=16 B,partitions=1,parallelTasks=3,preds=[(BETA.A[7:3] = subq=12)])
            ->  Subquery(n=12,totalCost=17.347,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=3,correlated=true,expression=true,invariant=true)
              ->  ProjectRestrict(n=12,totalCost=13.337,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=3)
                ->  GroupBy(n=11,totalCost=13.337,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=3)
                  ->  ProjectRestrict(n=10,totalCost=4.001,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3)
                    ->  TableScan[T2(1664)](n=8,totalCost=4.001,scannedRows=1,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3,preds=[(SUBQ.B[8:2] = ALPHA.B[7:2])])
            ->  MergeSortJoin(n=6,totalCost=8.008,outputRows=1,outputHeapSize=16 B,partitions=1,parallelTasks=3,preds=[(ALPHA.B[6:2] = BETA.B[6:4])])
              ->  TableScan[T2(1664)](n=4,totalCost=4.001,scannedRows=1,outputRows=1,outputHeapSize=16 B,partitions=1,parallelTasks=3)
              ->  TableScan[T1(1648)](n=2,totalCost=4.001,scannedRows=1,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3)
      ->  ProjectRestrict(n=1,totalCost=4.001,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3)
        ->  TableScan[T1(1648)](n=0,totalCost=4.001,scannedRows=1,outputRows=1,outputHeapSize=8 B,partitions=1,parallelTasks=3)

```